### PR TITLE
Handle large data responses

### PR DESCRIPTION
**Pull request by darraghgrace**

### CODE REVIEW

**Comment by lukeis** [comment]
:+1:

**Comment by darraghgrace** [comment]
@freynaud - Any objections to getting this merged?

**Comment by mach6** [comment]
I think we should merge this.

**Comment by darraghgrace** [comment]
@mmmulani @abhijit1234 @mach6 : guys, i see you are looking at PR's recently, can you review this, I would like to get it merged. Thanks

**Comment by mmmulani** [comment]
this looks mostly good but why have the `jsonMap`? it's only ever used for one session "uuid" and then wiped.
I think it'd be clearer if you just had a response that you kept building up and wiped after you received the final terminator

File: server/src/main/resources/instruments-js/UIAutomation.js [hunk 1]
```diff
@@ -98,11 +98,42 @@ var UIAutomation = {
...
+            var uuidLength = 36;
+            var jsonPart;
+            var index = 0;
+            var jsonRemainder = jsonResponse;
+            var startPos = jsonResponse.indexOf(sessionId) + sessionId.length + 3;
+            var sessionUUID = jsonResponse.substring(startPos, startPos + uuidLength);
+
+            while (jsonRemainder.length > maxPartLength){
+                jsonPart =  sessionUUID + jsonRemainder.substring(0,maxPartLength) + partialIdentifier;
+                jsonRemainder = jsonRemainder.substring(maxPartLength, Number.MAX_VALUE);
```
Comment (abhijit1234) [inline server/src/main/resources/instruments-js/UIAutomation.js:116]: [Optional]: You can just do

```
 jsonRemainder.substring(maxPartLength);
```

File: server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java [hunk 2]
```diff
@@ -112,7 +116,24 @@ private void getResponse(HttpServletRequest request, HttpServletResponse respons
...
-        String json = writer.toString();
+        String json = null;
+        String sessionUUID = null;
+        String jsonConcatenated = "";
+        if (writer.toString().endsWith(PARTIAL_IDENTIFIER)){
+            sessionUUID = writer.toString().substring(0, UUID_LENGTH);
+            if (jsonMap.get(sessionUUID) != null){
+                jsonConcatenated = (String)jsonMap.get(sessionUUID);
+            }
+            jsonConcatenated =  jsonConcatenated.concat(writer.toString().substring(UUID_LENGTH, writer.toString().length() - PARTIAL_IDENTIFIER.length()));
```
Comment (abhijit1234) [inline server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java:127]: Should it be this instead

```
substring(UUID_LENGTH, writer.toString().length() - PARTIAL_IDENTIFIER.length()-UUID_LENGTH)
```

File: server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java [hunk 3]
```diff
@@ -112,7 +116,24 @@ private void getResponse(HttpServletRequest request, HttpServletResponse respons
...
+            jsonMap.put(sessionUUID, jsonConcatenated);
+            return;
+        }else if (writer.toString().endsWith(FINAL_PART_IDENTIFIER)){
+            sessionUUID = writer.toString().substring(0, UUID_LENGTH);
+            json =  ((String)jsonMap.get(sessionUUID)).concat(writer.toString().substring(UUID_LENGTH, writer.toString().length() - FINAL_PART_IDENTIFIER.length()));
```
Comment (abhijit1234) [inline server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java:132]: Same comment as above

File: server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java (hunk 2 above)
Comment (acrehan) [inline server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java:127]: This suggested change will break the fix as it is chopping off an extra 36 chars.

File: server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java (hunk 3 above)
Comment (acrehan) [inline server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java:132]: This suggested change will break the fix as it is chopping off an extra 36 chars.

**Comment by acrehan** [comment]
> > this looks mostly good but why have the jsonMap? 
> > The jsonMap is needed to handle the case of multiple concurrent sessions.

File: server/src/main/resources/instruments-js/UIAutomation.js (hunk 1 above)
Comment (acrehan) [inline server/src/main/resources/instruments-js/UIAutomation.js:116]: Agreed. Changing as suggested thanks.

File: server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java (hunk 2 above)
Comment (abhijit1234) [inline server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java:127]: You are constructing the json parts as

```
Each json part except last part = <36 character session id><Some part of the original response><44 character partial id>
Last json part                  = <36 character session id><Some part of the original response><44 character final id>
```

let's say 

```
writer.toString().length() is equal to n
len(PARTIAL_IDENTIFIER) = 44
```

so you are doing

```
jsonConcatenated =  jsonConcatenated.concat(writer.toString().substring(36, n - 44));
```

which implies 

```
jsonConcatenated = <part 1 of original response>b94b2b80-4330-11e4-916c-0800200c9a66<part 2 of original response>b94b2b80-4330-11e4-916c-0800200c9a66...<part n of original response>b94b2b80-4330-11e4-916c-0800200c9a66>
```

I believe what you want is

```
jsonConcatenated = <part 1 of original response><part 2 of original response>...<part n of original response>b94b2b80-4330-11e4-916c-0800200c9a66>
```

Comment (acrehan) [inline server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java:127]: > > I believe what you want is
> > jsonConcatenated = <part 1 of original response><part 2 of original response>...<part n of original >>response>b94b2b80-4330-11e4-916c-0800200c9a66>.
> > This is what I am currently getting.

Comment (abhijit1234) [inline server/src/main/java/org/uiautomation/ios/instruments/commandExecutor/CURLIAutomationCommandExecutor.java:127]: Sorry for the confusion. I misread javadocs. Seems fine to me

**Comment by mmmulani** [comment]
makes sense, didn't realise the concurrent use case

**Pull request merged**